traffic_ops/traffic_ops_golang/federations/allfederations.go

```diff
@@ -0,0 +1,132 @@
+package federations
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import (
+	"database/sql"
+	"errors"
+	"net/http"
+
+	"github.com/apache/trafficcontrol/lib/go-tc"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/api"
+)
+
+func GetAll(w http.ResponseWriter, r *http.Request) {
+	inf, userErr, sysErr, errCode := api.NewInfo(r, nil, nil)
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+	defer inf.Close()
+
+	feds := []FedInfo{}
+	err := error(nil)
+	allFederations := []tc.IAllFederation{}
+
+	if cdnParam, ok := inf.Params["cdnName"]; ok {
+		cdnName := tc.CDNName(cdnParam)
+		feds, err = getAllFederationsForCDN(inf.Tx.Tx, cdnName)
+		if err != nil {
+			api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New("federations.GetAll getting all federations: "+err.Error()))
+			return
+		}
+		allFederations = append(allFederations, tc.AllFederationCDN{CDNName: &cdnName})
```

Comment:
I honestly don't know, I'd have to dive into the Router. I just reproduced the existing endpoint, because I didn't want to spend the time diving into the Router code, or changing it.

I'd certainly support changing it, it's awkward and heterogeneous JSON objects are a pain.

We should be able to remove it from `api/1.4`, any time before 4.0 is cut.

Reply:
Yeah, so TR accepts a configurable URL via its profile, but by default it points to that internal endpoint without `?cdnName=foo`. If we ever _did_ want to add that qstring to the URL, we'd have to fix this API to not return that `tc.AllFederationCDN` object. I'll open an issue after merging so we can track that work.